tap_dynamodb/sync_strategies/log_based.py

```diff
@@ -174,63 +165,69 @@ def sync_log_based(config, state, stream):
                                       table_name, stream_version, state)
 
-        # If the shard we just finished syncing is closed (i.e. has an
-        # EndingSequenceNumber), remove it from the seq_number_bookmark
-        # and put it into the finished_shards bookmark
-        if shard['SequenceNumberRange'].get('EndingSequenceNumber'):
-            # Must check if the bookmark exists because if a shard has 0
-            # records we will never set a bookmark for the shard
-            if seq_number_bookmarks.get(shard['ShardId']):
-                seq_number_bookmarks.pop(shard['ShardId'])
-                finished_shard_bookmarks.append(shard['ShardId'])
-                state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
-                state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
-                singer.write_state(state)
-
-    for shard in finished_shard_bookmarks:
-        if shard not in found_shards:
+        # Now that we have fully synced the shard, move it from the
+        # shard_seq_numbers to finished_shards.
+        finished_shard_bookmarks.append(shard['ShardId'])
+        state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
+
+        if seq_number_bookmarks.get(shard['ShardId']):
+            seq_number_bookmarks.remove(shard['ShardId'])
+            state = singer.write_bookmark(state, table_name, 'shard_seq_numbers', seq_number_bookmarks)
+
+        singer.write_state(state)
+
+    for shardId in finished_shard_bookmarks:
+        if shardId not in found_shards:
             # Remove this shard because its no longer appearing when we query for get_shards
-            finished_shard_bookmarks.remove(shard)
+            finished_shard_bookmarks.remove(shardId)
             state = singer.write_bookmark(state, table_name, 'finished_shards', finished_shard_bookmarks)
 
     singer.write_state(state)
 
     return rows_synced
 
-def has_stream_aged_out(config, state, stream):
+
+def has_stream_aged_out(state, table_name):
+    '''
+    Uses the success_timestamp on the stream to determine if we have
+    successfully synced the stream in the last 19 hours 30 minutes.
+
+    This uses 19 hours and 30 minutes because records are guaranteed to be
+    available on a shard for 24 hours, but we only process closed shards
+    and shards close after 4 hours. 24-4 = 20 and we gave 30 minutes of
+    wiggle room because I don't trust AWS.
+    '''
+    current_time = singer.utils.now()
+
+    success_timestamp = singer.get_bookmark(state, table_name, 'success_timestamp')
+
+    # If we have no success_timestamp then we have aged out
+    if not success_timestamp:
+        return True
+
+    time_span = current_time - singer.utils.strptime_to_utc(success_timestamp)
+
+    # If it has been > than 19h30m since the last successful sync of this
```

Comment:
Maybe we should link here to documentation that streams only ever holds 24 hours of data and after that shards are reaped / aged out?

Reply:
And that's why we do 19:30 because a shard stays open for 4 hours, hence 24 - 4 with a 30 minute buffer.

Reply:
Added a link to the python docstring at the top about this.